errors.go

```diff
@@ -2,120 +2,48 @@ package jwt
-	ValidationErrorIssuer        // ISS validation failed
-	ValidationErrorSubject       // SUB validation failed
-	ValidationErrorNotValidYet   // NBF validation failed
-	ValidationErrorId            // JTI validation failed
-	ValidationErrorClaimsInvalid // Generic claims validation error
+	ErrInvalidKey                = errors.New("key is invalid")
+	ErrInvalidKeyType            = errors.New("key is of invalid type")
+	ErrHashUnavailable           = errors.New("the requested hash function is unavailable")
+	ErrTokenMalformed            = errors.New("token is malformed")
+	ErrTokenUnverifiable         = errors.New("token is unverifiable")
+	ErrTokenSignatureInvalid     = errors.New("token signature is invalid")
+	ErrTokenRequiredClaimMissing = errors.New("token is missing required claim")
+	ErrTokenInvalidAudience      = errors.New("token has invalid audience")
+	ErrTokenExpired              = errors.New("token is expired")
+	ErrTokenUsedBeforeIssued     = errors.New("token used before issued")
+	ErrTokenInvalidIssuer        = errors.New("token has invalid issuer")
+	ErrTokenInvalidSubject       = errors.New("token has invalid subject")
+	ErrTokenNotValidYet          = errors.New("token is not valid yet")
+	ErrTokenInvalidId            = errors.New("token has invalid id")
+	ErrTokenInvalidClaims        = errors.New("token has invalid claims")
+	ErrInvalidType               = errors.New("invalid type for claim")
 )
 
-// NewValidationError is a helper for constructing a ValidationError with a string error message
-func NewValidationError(errorText string, errorFlags uint32) *ValidationError {
-	return &ValidationError{
-		text:   errorText,
-		Errors: errorFlags,
-	}
+// joinedError is an error type that works similar to what [errors.Join]
+// produces, with the exception that it has a nice error string; mainly its
+// error messages are concatenated using a comma, rather than a newline.
+type joinedError struct {
+	errs []error
 }
 
-// ValidationError represents an error from Parse if token is not valid
-type ValidationError struct {
-	// Inner stores the error returned by external dependencies, e.g.: KeyFunc
-	Inner error
-	// Errors is a bit-field. See ValidationError... constants
-	Errors uint32
-	// Text can be used for errors that do not have a valid error just have text
-	text string
-}
-
-// Error is the implementation of the err interface.
-func (e ValidationError) Error() string {
-	if e.Inner != nil {
-		return e.Inner.Error()
-	} else if e.text != "" {
-		return e.text
-	} else {
-		return "token is invalid"
+func (je joinedError) Error() string {
+	msg := []string{}
+	for _, err := range je.errs {
+		msg = append(msg, err.Error())
 	}
-}
-
-// Unwrap gives errors.Is and errors.As access to the inner error.
-func (e *ValidationError) Unwrap() error {
-	return e.Inner
-}
 
-// No errors
-func (e *ValidationError) valid() bool {
-	return e.Errors == 0
+	return strings.Join(msg, ", ")
```

Comment:
```suggestion
	return strings.Join(msg, ": ")
```

Would it be more idiomatic if these were separated by colons (`:`) ?

No strong preference, simply an observation that colon is more widely used (at least based on my experience and those who adopted pkg/errors).

Reply:
Errors that are caused by another are concatenated using `:`, errors that are sort of "in parallel" are separated by `,` So that basically you end up with something that reads: 
```
token has invalid claims: token is expired, token is not valid yet
```

If we choose to only use `:` you would end up with 

```
token has invalid claims: token is expired: token is not valid yet
```

which would be weird, since the "not valid yet" is not a cause of "expired", but basically exists next to each other.

The new Go 1.20 `errors.Join` uses a newline to join them; we can also simulate that but this is a bit weird I think personally: 
```
token has invalid claims: token is expired
token not valid yet
```

Reply:
Okay, that's fair. I agree the newline is an odd choice and the inlined version (as you have it) is better.

The newline assumes it's always being printed out, but that's unlikely to be the case. Most often I wager these will end up in some structured logger and the newline will make it harder to read.